crates/worker/src/partition/snapshots/repository.rs

```diff
@@ -0,0 +1,307 @@
+    ) -> anyhow::Result<SnapshotRepository> {
+        let destination = if let Some(ref destination) = snapshots_options.destination {
+            destination.clone()
+        } else {
+            base_dir
+                .join("pp-snapshots")
+                .into_os_string()
+                .into_string()
+                .map(|path| format!("file://{path}"))
+                .map_err(|e| anyhow!("Unable to convert path to string: {:?}", e))?
+        };
+        let destination =
+            Url::parse(&destination).context("Failed parsing snapshot repository URL")?;
+
+        // AWS-specific ergonomics optimization: without explicit configuration, we set up the AWS
+        // SDK credentials provider so that the conventional environment variables and config
+        // locations just work. This makes object_store behave similarly to the Lambda invoker.
+        let object_store: Arc<dyn ObjectStore> = if destination.scheme() == "s3"
+            && destination.query().is_none()
+            && snapshots_options.additional_options.is_empty()
+        {
+            debug!("Using AWS SDK credentials provider");
+            let aws_region = aws_config::load_defaults(BehaviorVersion::v2024_03_28())
+                .await
+                .region()
+                .context("Unable to determine AWS region to use with S3")?
+                .clone();
+
+            let store = AmazonS3Builder::new()
+                .with_url(destination.clone())
+                .with_region(aws_region.to_string())
+                .with_credentials(Arc::new(AwsSdkCredentialsProvider {
+                    credentials_provider: DefaultCredentialsChain::builder().build().await,
+                }))
+                .build()?;
+
+            Arc::new(store)
+        } else {
+            debug!("Using object_store credentials configuration");
+            object_store::parse_url_opts(&destination, &snapshots_options.additional_options)?
+                .0
+                .into()
+        };
+
+        // prefix must be stripped of any leading slash and, unless zero-length, end in a single "/" character
+        let prefix: String = destination.path().into();
+        let prefix = match prefix.as_str() {
+            "" | "/" => "".to_string(),
+            prefix => format!("{}/", prefix.trim_start_matches('/').trim_end_matches('/')),
+        };
+
+        Ok(SnapshotRepository {
+            object_store,
+            destination,
+            prefix,
+        })
+    }
+
+    /// Write a partition snapshot to the snapshot repository.
+    pub(crate) async fn put(
+        &self,
+        snapshot: &PartitionSnapshotMetadata,
+        local_snapshot_path: PathBuf,
+    ) -> anyhow::Result<()> {
+        let partition_id = snapshot.partition_id;
+        let lsn = snapshot.min_applied_lsn;
+
+        debug!(
+            %lsn,
+            "Publishing partition snapshot to: {}",
+            self.destination,
+        );
```

Comment:
You can instrument `put` via `#[instrument()]` and include the `lsn`, snapshot id, etc.